### ISSUE TITLE
Azure queue stream provider data loss.

### ISSUE THREAD

**Issue by jason-bragg** [dotnet/orleans#1615]
Since azure queues are unordered queues, we add ordering information to the events derived from the order in which we read them from the queue.  This ordering is used by the cache and stream grain extensions to deliver the correct data to a grain in the correct order.  When an azure queue is moved between silos (or shutdown and restarted for any reason) the ordering information in the events is reset.  Any active grains, expecting events with ordering information consistent with the previous numbering, will not receive events until the new ordering information catches up.  Until this occurs, all messages will be lost for that consumer.

Example:
Azure queue has a single stream of 300 events.
Queue is read on silo S1
Grain G1 subscribes
First 100 events are sent to grain ordered as 1, 2, 3 ... 100.
Queue is moved to silo S2
Next 200 events are sent to grain in order of 1, 2, 3, ... 200.
Since grain has received events with order information up to 100, it doesn't get the first 100 events.
So out of the 300 events in the stream G1 will only see 200: 1-100, 200-300.

StreamReliabilityTests.AQ_StreamRel_SiloJoins test fails intermittently because of this issue.

**Comment by gabikliot** [comment]
@jason-bragg , I had to read your description 3 times until I think I got it.
Lets see if I got it correctly.

1) AzureQueue does NOT support ordering. 
2) That means Azure Queue Stream Provider should not even try to support ordering.
3) Application using Azure Queue Stream Provide should not expect ordered delivery.
4) Our implementation of Azure Queue Stream Provider actually tries (erroneously) to provide some order for AQSP, fails at that (basically since it is not even supposed to), but as a result there is a bug in the impl. that also prevents some events to arrive.

What confused me in our description is:
>This ordering is used by the cache and stream grain extensions to deliver the correct data to a grain in the correct order

This is not correct. The ordering is used, to help cache management, but it SHOULD NOT be used "to deliver the correct data to a grain in the correct order", since as said above, in AQSP case, there is no such thing as correct order, by definition on AQ being un-ordered.

Now I looked at the code and I think I clearly see the bug.
The application (the grain code in `StreamReliabilityTests.AQ_StreamRel_SiloJoins`) correctly does NOT expect any ordering.
But we (you and me together) added ordering a while ago in the extension implementation, to provide ordering for Event Hub provider (which does guarantee ordering). What we forgot to do is to disable ordering check in AQ case.
Basically, this line:
https://github.com/dotnet/orleans/blob/master/src/Orleans/Streams/Internal/StreamSubscriptionHandleImpl.cs#L73
should not be executed for AQ stream provider.

Do you agree?

**Comment by gabikliot** [comment]
I implemented a fix in #1616.

**Comment by jason-bragg** [comment]
> This ordering is used by the cache and stream grain extensions to deliver the correct data to a grain in the correct order - "This is not correct."

This was a statement of what the system does, as part of a description of the problem, not what it should do.

I don't think this is as simple as not correcting out of order data, it is more related to where in the cache the agent starts to deliver from after moving a queue.  When an agent sets up a subscription to a grain, it performs a handshake to determine where to start delivering data.  If this is being done by a new agent to an existing grain, the grain will return a token that appears to be far in the future because the ordering in the cache was reset with the queue relocation.  This disparity is the primary cause of data loss.

As unordered data is less strict than ordered data, my view was that if the streaming infrastructure handled ordered data, then it would also handle data coming from unordered queues if we could treat the data as ordered where the order is the order in which it comes off of the queue.  This mostly works, except when we move the queue and the ordering is reset (this bug).  I remain of the opinion that this approach is simpler and less error prone than having a streaming infrastructure that must function under multiple delivery modes.  If we continue under this model then the issue becomes:

"How do we ensure data ordering from newer agents is always higher than those from older agents?" 

We could use time, or some seed from the silo stats/membership protocol..?

The alternative approach is to require the streaming infrastructure to support an ordered/unordered mode.  This is possible but introduces complexity and nuance that makes the system harder to test and maintain.

A more extreme, but possibly better approach, may be to support different agents (and possibly grain extensions) for ordered and unordered data.  While this would be more code, I suspect the code would be more testable and maintainable.

**Comment by gabikliot** [comment]
Lets look at my impl. again, more carefully.

> When an agent sets up a subscription to a grain, it performs a handshake to determine where to start delivering data. If this is being done by a new agent to an existing grain, the grain will return a token that appears to be far in the future because the ordering in the cache was reset with the queue relocation. This disparity is the primary cause of data loss.

In my fix, the grain always returns null as expected token (`expectedToken` is never set if `!isRewindable`), so it stays null always. The agent in such case will just start from the beginning of the cache.
Am I wrong?

Lets first see if this fix can (easily, as done now) fix the issue and then discuss alternative solutions, what do you think?

**Comment by jason-bragg** [comment]
I was mistaken.  I am now also of the opinion that you're impl addresses the issue.  I simply didn't think through it sufficiently. :/
That being said, I believe the fix can be simplified a little.  Plz see my comments in #1616.

**Comment by gabikliot** [comment]
:+1:

**Comment by jason-bragg** [comment]
Fixed with  #1616
